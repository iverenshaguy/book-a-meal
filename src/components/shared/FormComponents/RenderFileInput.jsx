import React, { Component } from 'react';
import PropTypes from 'prop-types';
import MiniPreloader from '../Preloader/MiniPreloader';
import fileEventAdapter from '../../../utils/fileEventAdapter';
import uploadValidation from '../../../helpers/validations/uploadValidation';

/**
 * @exports
 * @class RenderFileInput
 * @extends Component
 * @return {JSX} RenderFileInput
 */
class RenderFileInput extends Component {
  /**
   * @constructor
   * @return {JSX} RenderFileInput
   */
  constructor() {
    super();

    this.state = {
      error: null
    };
  }

  /**
   * @memberof Form
   * @returns {void}
  */
  handleClick = () => {
    const { clearUploadError } = this.props;

    this.setState({ error: null });

    clearUploadError();
  }

  /**
   * @memberof Form
   * @param {object} event
   * @param {element} preview
   * @returns {void}
   */
  handleChange = (event) => {
    const file = event.target.files[0];
    const maxSize = 2 * 1024 * 1024; // 2MB max size
    const allowedTypes = ['image/gif', 'image/jpeg', 'image/png'];
    const { formerImgUrl, successCallBack, uploadImage } = this.props;
    const preview = document.querySelector('#meal-form-image');

    if (!uploadValidation(file, maxSize, allowedTypes)) {
      fileEventAdapter(preview)(event);
      const fileType = file.type;
      const fileExt = fileType.split('/')[1];

      return uploadImage(file, formerImgUrl, `images/${Date.now()}.${fileExt}`, successCallBack);
    }

    this.setState({ error: uploadValidation(file, maxSize, allowedTypes) });
    // reset input box
    event.target.value = '';
  }

  /**
   * @function RenderFileInput
   * @return {JSX} RenderFileInput
   */
  render() {
    const {
      id, name, label, uploading, uploadError, updating, formerImgUrl
    } = this.props;
    const error = this.state.error || uploadError;

    return (
      <div className="form-input form-img-input">
        {label && <label htmlFor={id}>{label}</label>}
        <div id="form-image-preview">
          {(uploading || updating) && <div className="img-overlay" id="img-overlay"><MiniPreloader /></div>}
          <img src={formerImgUrl} alt="meal" id="meal-form-image" />
        </div>
        <input
          id={id}
          type="file"
          name={name}
          accept="image/*"
          onClick={this.handleClick}
          onChange={this.handleChange}
        />
        <br />
        {error && <small className="file-feedback danger">{error}</small>}
      </div>
    );
  }
}

RenderFileInput.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  updating: PropTypes.bool,
  uploading: PropTypes.bool.isRequired,
  uploadImage: PropTypes.func.isRequired,
  uploadError: PropTypes.string,
  formerImgUrl: PropTypes.string.isRequired,
  clearUploadError: PropTypes.func.isRequired,
  successCallBack: PropTypes.func.isRequired,
};

RenderFileInput.defaultProps = {
  uploadError: null,
  updating: null
};

export default RenderFileInput;
