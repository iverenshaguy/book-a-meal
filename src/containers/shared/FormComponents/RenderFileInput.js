import { connect } from 'react-redux';
import RenderFileInput from '../../../components/shared/FormComponents/RenderFileInput';
import { uploadImage, clearUploadError } from '../../../actions/uploadImage';

const mapStateToProps = state => ({
  uploadError: state.uploadImage.error,
  uploading: state.uploadImage.uploading,
});

export default connect(mapStateToProps, { uploadImage, clearUploadError })(RenderFileInput);
