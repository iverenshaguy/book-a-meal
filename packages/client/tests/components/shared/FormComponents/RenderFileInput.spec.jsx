import React from 'react';
import RenderFileInput from '../../../../src/components/shared/FormComponents/RenderFileInput';

const props = {
  id: '1',
  name: 'image',
  label: 'Image',
  updating: false,
  uploading: false,
  uploadImage: jest.fn(),
  uploadError: null,
  formerImgUrl: 'image.jpg',
  clearUploadError: jest.fn(),
  successCallBack: jest.fn(),
};

const goodFile = new Blob([''], {
  type: 'image/jpeg',
  size: '500'
});

const badFile = new Blob([''], {
  type: 'image/pdf',
  size: '500'
});

describe('Form Components: RenderFileInput', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<RenderFileInput {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('div.file-feedback').exists()).toBeFalsy();
  });

  it('should show error when there is an upload error', () => {
    const wrapper = shallow(<RenderFileInput {...props} uploadError="This is an error" />);

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('.file-feedback').text()).toEqual('This is an error');
  });

  it('should show preloader when uploading', () => {
    const wrapper = shallow(<RenderFileInput {...props} uploading />);

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('MiniPreloader')).toBeTruthy();
  });

  it('should call handle click on input click', () => {
    const uploadImageMock = jest.fn();
    const wrapper = shallow(<RenderFileInput {...props} uploadImage={uploadImageMock} />);

    wrapper.setState({ error: 'error' });
    wrapper.find('input').simulate('click');

    expect(wrapper.state().error).toEqual(null);
  });

  it('should call uploadImage if a valid file is passed in and there is no upload error', () => {
    const uploadImageMock = jest.fn();
    const wrapper = shallow(<RenderFileInput {...props} uploadImage={uploadImageMock} />);
    const event = {
      target: {
        files: [goodFile]
      }
    };

    wrapper.find('input').simulate('change', event);

    expect(uploadImageMock).toHaveBeenCalled();
  });

  it('should not call uploadImage if an invalid file is passed in', () => {
    const uploadImageMock = jest.fn();
    const wrapper = shallow(<RenderFileInput {...props} uploadImage={uploadImageMock} />);
    const event = {
      target: {
        files: [badFile]
      }
    };

    wrapper.find('input').simulate('change', event);

    expect(uploadImageMock).not.toHaveBeenCalled();
  });
});
