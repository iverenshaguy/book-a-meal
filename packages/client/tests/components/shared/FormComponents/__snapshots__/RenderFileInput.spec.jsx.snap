// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Components: RenderFileInput should render correctly 1`] = `
<div
  className="form-input form-img-input"
>
  <label
    htmlFor="1"
  >
    Image
  </label>
  <div
    id="form-image-preview"
  >
    <img
      alt="meal"
      id="meal-form-image"
      src="image.jpg"
    />
  </div>
  <input
    accept="image/*"
    id="1"
    name="image"
    onChange={[Function]}
    onClick={[Function]}
    type="file"
  />
  <br />
</div>
`;

exports[`Form Components: RenderFileInput should show error when there is an upload error 1`] = `
<div
  className="form-input form-img-input"
>
  <label
    htmlFor="1"
  >
    Image
  </label>
  <div
    id="form-image-preview"
  >
    <img
      alt="meal"
      id="meal-form-image"
      src="image.jpg"
    />
  </div>
  <input
    accept="image/*"
    id="1"
    name="image"
    onChange={[Function]}
    onClick={[Function]}
    type="file"
  />
  <br />
  <small
    className="file-feedback danger"
  >
    This is an error
  </small>
</div>
`;

exports[`Form Components: RenderFileInput should show preloader when uploading 1`] = `
<div
  className="form-input form-img-input"
>
  <label
    htmlFor="1"
  >
    Image
  </label>
  <div
    id="form-image-preview"
  >
    <div
      className="img-overlay"
      id="img-overlay"
    >
      <MiniPreloader />
    </div>
    <img
      alt="meal"
      id="meal-form-image"
      src="image.jpg"
    />
  </div>
  <input
    accept="image/*"
    id="1"
    name="image"
    onChange={[Function]}
    onClick={[Function]}
    type="file"
  />
  <br />
</div>
`;
